public/app/core/components/Login/ChangePassword.tsx

```diff
@@ -0,0 +1,126 @@
+import React, { PureComponent, SyntheticEvent } from 'react';
+import { Tooltip } from '@grafana/ui';
+import appEvents from 'app/core/app_events';
+
+interface Props {
+  onSubmit: (pw: { newPassword: string; confirmNew: string; oldPassword: string }, valid: boolean) => void;
+  onSkip: Function;
+  className?: string;
+  focus?: boolean;
+}
+
+interface State {
+  newPassword: string;
+  confirmNew: string;
+  valid: boolean;
+}
+
+export class ChangePassword extends PureComponent<Props, State> {
+  private userInput: HTMLInputElement;
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      newPassword: '',
+      confirmNew: '',
+      valid: false,
+    };
+  }
+
+  componentDidUpdate(prevProps: Props) {
+    if (!prevProps.focus && this.props.focus) {
+      this.focus();
+    }
+  }
+
+  focus() {
+    this.userInput.focus();
+  }
+
+  onSubmit = (e: SyntheticEvent) => {
+    e.preventDefault();
+
+    const { newPassword, confirmNew } = this.state;
+    if (this.state.valid) {
+      this.props.onSubmit({ newPassword, confirmNew, oldPassword: 'admin' }, true);
+    } else {
+      appEvents.on('alert-warning', ['New passwords do not match', '']);
+    }
+  };
+
+  onChange = (e: SyntheticEvent) => {
+    // @ts-ignore
+    this.setState({ [e.target.name]: e.target.value }, () => {
+      this.setState({
```

Comment:
Seems like an antipattern to me to set state as a consequence of setState (will cause double re-render). I think you should validate the data before setting new value.

Reply:
Are you thinking that I should only set the values if they are validated? Shouldn't the state reflect whatever is in the text field?

Reply:
it should, but you have the knowledge what the text gonna be as it's available via the event object, right? sou you could do :

```ts
this.setState((state) => {
  [e.target.name]: e.target.value,
  isValid: /// compare the event value against the state value
});
```